src/components/MachineSelector/Configurator/MachineConfigurator.tsx

```diff
@@ -0,0 +1,337 @@
+    props.setAlerts({
+      type: 'error',
+      text: msg,
+      hideApp: false,
+    });
+  };
+
+  const { handleSubmit, control, reset } = useForm({
+    defaultValues,
+  });
+
+  /**
+   * This will open the configurator popup on machine settings icon click.
+   * restoreMachineConfig function is used to fill the form fields with previously saved values
+   */
+  const handleClickOpen = () => {
+    setOpen(true);
+    setTimeout(() => {
+      restoreMachineConfig();
+    }, 10);
+  };
+
+  /**
+   * This will close the popup
+   */
+  const handleClose = () => {
+    setOpen(false);
+  };
+
+  /**
+   * This is used to update the form fields with previously saved values
+   */
+  const restoreMachineConfig = async () => {
+    const savedMachineConfigStr = await getMachineIds(onError);
+    if (savedMachineConfigStr) {
+      setAssetExternalIds(savedMachineConfigStr.split(','));
+    }
+  };
+
+  /**
+   * This function is used to validate machine saving.
+   */
+  const isValidSaving = () => {
+    if (assetExternalIds.length === 0) {
+      props.setAlerts({
+        type: 'warning',
+        text: 'Atleast one Asset External Id is required!',
+        hideApp: false,
+      });
+      return false;
+    }
+    return true;
+  };
+
+  /**
+   * This function is used to save machine configurations.
+   * A alert will be displayed finally based on the saving status.
+   */
+  const saveMachineConfig = async () => {
+    if (isValidSaving()) {
+      const actionStatus = await saveMachineIds(
+        assetExternalIds.join(','),
+        onError
+      );
+      if (actionStatus) {
+        handleClose();
+        props.updateAssets();
+        props.setAlerts({
+          type: 'success',
+          text: 'Successfully saved!',
+          hideApp: false,
+        });
+      }
+    }
+  };
+
+  /**
+   * This method is used to remove asset id from the list
+   */
+  const removeAssetId = (id: string) => {
+    const idList = assetExternalIds.filter(
+      assetExternalId => assetExternalId !== id
+    );
+    setAssetExternalIds(idList);
+  };
+
+  /**
+   * This method is used to add asset id to the list
+   */
+  const addAssetId = (assetId: string) => {
+    setAssetExternalIds([assetId, ...assetExternalIds]);
+    reset();
+  };
+
+  /**
+   * This method is used to validate asset id before it adding to the list
+   */
+  const validateAssetId = async (assetId: string) => {
+    if (assetExternalIds.includes(assetId)) {
+      onError('Asset External Id is already added!');
+    } else {
+      setAssetValidating(true);
+      const assetIdStatus = await validateAsset(assetId);
+      setAssetValidating(false);
+      if (assetIdStatus) {
+        addAssetId(assetId);
+      } else {
+        onError('Asset not found!');
+      }
+    }
+  };
+
+  /**
+   * This function fires on form submit
+   */
+  const onSubmit = (data: FormField) => {
+    validateAssetId(data.assetExternalId);
+  };
+
+  return (
```

Comment:
Some comments at-least describing different parts would be helpful.

Reply:
is this for html(JSX) code?

Reply:
Done